In [1]:
import torch
import gc
torch.cuda.empty_cache()
gc.collect()


0

In [2]:
from setup import setup_src_path
print(setup_src_path())
import data.processed as processed
import config.config as config
import utils.setup as setup
import utils.functions as fn
from importlib import reload



['/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/mlm', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '', '/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages', '/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/modules']


2024-06-05 00:23:48.253468: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-06-05 00:23:48.455910: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-06-05 00:23:49.330115: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [3]:
from datasets import load_from_disk,concatenate_datasets


adapter_name="mlm_union_all_G"
dataset = load_from_disk(f"{config.Config.DATASETS_SAVE_PATH}/datasets")



In [4]:
reload(config)

    
filtered_target = dataset['train'].filter(lambda example: example['genre'] == "government").shuffle(seed=42)
#filtered_target = shuffled_filtered_target.train_test_split(test_size=0.1)


filtered_test_target = dataset['validation_matched'].filter(lambda example: example['genre'] == "government")
train_target = filtered_target
test_target = filtered_test_target

 

In [5]:
from adapters import AutoAdapterModel,init
from tqdm import tqdm

from transformers import AutoConfig, AutoModelForMaskedLM

mdlcfg = AutoConfig.from_pretrained(
    config.Config.MODEL_NAME,
 
)
model = AutoModelForMaskedLM.from_pretrained(
    config.Config.MODEL_NAME,
)
init(model)
reload(fn)
fn.print_trainable_parameters(model)

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertFor

trainable params: 109514298 || all params: 109514298 || trainable%: 100.0


In [6]:
from adapters import ConfigUnion, LoRAConfig, PrefixTuningConfig, SeqBnConfig,SeqBnInvConfig,AdapterConfig,LoRAConfig

union_config = ConfigUnion(
    PrefixTuningConfig(prefix_length=10),
    LoRAConfig(r=8, alpha=16),

    SeqBnInvConfig(reduction_factor=2),
)
model.add_adapter(adapter_name, config=union_config)

model.train_adapter([adapter_name])
model.active_adapters = adapter_name
print(model.adapter_summary())
fn.print_trainable_parameters(model)

Name                     Architecture         #Param      %Param  Active   Train
--------------------------------------------------------------------------------
mlm_union_all_G          union            17,539,712      16.107       1       1
--------------------------------------------------------------------------------
Full model                               108,891,648     100.000               0
trainable params: 18131840 || all params: 127054010 || trainable%: 14.2709702747674


In [7]:
from transformers import AutoTokenizer

reload(processed)
tokenizer = AutoTokenizer.from_pretrained(config.Config.TOKENIZER_NAME)

tokenized_dataset= processed.tokenize_dataset(train_target,tokenizer)
tokenized_dataset

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids'],
    num_rows: 77350
})

In [8]:
chunk_size = 128
# Slicing produces a list of lists for each feature
tokenized_samples = tokenized_dataset[444:470]

results = fn.group_texts(tokenized_samples, chunk_size)
for chunk in results["labels"]:
    print(f"'>>> Chunk length: {len(chunk)}'")

'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'


In [9]:
# we group texts and chunk them
lm_datasets = tokenized_dataset.map(fn.group_texts, batched=True,fn_kwargs={'chunk_size': chunk_size})
lm_datasets

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
    num_rows: 25770
})

In [10]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
samples = [lm_datasets[i] for i in range(2)]
for sample in samples:
    _ = sample.pop("word_ids")
dd=data_collator(samples)
for chunk in dd["input_ids"]:
    print(f"\n'>>> {tokenizer.decode(chunk)}'")


'>>> [CLS] some investments in stewardship [MASK], [MASK] example [MASK] parks, will libyan reported by both 1 [MASK] [MASK] [MASK] of bergen used as a park or an historic site [MASK] the steward wandered land category, and 2 ) by the number of physical units identified as national parks in the heritage assets [MASK] [MASK] [SEP] the number of acres used and the number of physical [MASK] gaa on [MASK]ship land are both reported. [SEP] [CLS] the benefits to be gained in air quality were calculated [MASK] epa for the first six years ( including the interim [MASK] ) [MASK] [MASK] reductions of [MASK] [MASK] 000 tons of hydrocarbons, 2, 388, 000 tons [MASK] [MASK] monoxide'

'>>> and 450, 000 tons of nitrogen oxide. [SEP] epa estimate [MASK], in the [MASK] six years, there will be 125, 000 fewer tons of [MASK]carbons. [SEP] [CLS] 6 trillion [MASK] [SEP] 10 trillion [SEP] [CLS] professional training organizations to offer on - campus [MASK] at very [SEP] there will now be [MASK] places [MA

In [11]:

downsampled_dataset = lm_datasets.train_test_split(
    test_size=0.1, seed=42
)
downsampled_dataset

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 23193
    })
    test: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 2577
    })
})

In [12]:

reload(fn)
trainer = fn.train_mlm_model(model,adapter_name,data_collator,tokenizer, downsampled_dataset['train'],downsampled_dataset['test'])

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/accelerate/accelerator.py:446: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(


In [13]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/81 [00:00<?, ?it/s]

>>> Perplexity: 16.83


In [14]:
trainer.train()

  0%|          | 0/14500 [00:00<?, ?it/s]

{'loss': 2.2929, 'grad_norm': 1.9824577569961548, 'learning_rate': 9.84e-05, 'epoch': 1.0}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.9731937646865845, 'eval_runtime': 4.5789, 'eval_samples_per_second': 562.802, 'eval_steps_per_second': 17.69, 'epoch': 1.0}
{'loss': 2.016, 'grad_norm': 1.7795733213424683, 'learning_rate': 9.322857142857144e-05, 'epoch': 2.0}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.818129301071167, 'eval_runtime': 4.6005, 'eval_samples_per_second': 560.154, 'eval_steps_per_second': 17.607, 'epoch': 2.0}
{'loss': 1.9266, 'grad_norm': 1.6789743900299072, 'learning_rate': 8.805714285714286e-05, 'epoch': 3.0}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.77144193649292, 'eval_runtime': 4.5989, 'eval_samples_per_second': 560.356, 'eval_steps_per_second': 17.613, 'epoch': 3.0}
{'loss': 1.8672, 'grad_norm': 1.6291940212249756, 'learning_rate': 8.289285714285714e-05, 'epoch': 3.99}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.7656924724578857, 'eval_runtime': 4.6577, 'eval_samples_per_second': 553.275, 'eval_steps_per_second': 17.39, 'epoch': 4.0}
{'loss': 1.8321, 'grad_norm': 1.7380824089050293, 'learning_rate': 7.772142857142857e-05, 'epoch': 4.99}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.7175638675689697, 'eval_runtime': 4.6043, 'eval_samples_per_second': 559.693, 'eval_steps_per_second': 17.592, 'epoch': 5.0}
{'loss': 1.8005, 'grad_norm': 1.6777958869934082, 'learning_rate': 7.255e-05, 'epoch': 5.99}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6983721256256104, 'eval_runtime': 4.6109, 'eval_samples_per_second': 558.892, 'eval_steps_per_second': 17.567, 'epoch': 6.0}
{'loss': 1.7797, 'grad_norm': 1.7608320713043213, 'learning_rate': 6.737857142857143e-05, 'epoch': 6.99}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6902989149093628, 'eval_runtime': 4.6326, 'eval_samples_per_second': 556.277, 'eval_steps_per_second': 17.485, 'epoch': 7.0}
{'loss': 1.7637, 'grad_norm': 1.7225887775421143, 'learning_rate': 6.220714285714286e-05, 'epoch': 7.99}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6575270891189575, 'eval_runtime': 4.5895, 'eval_samples_per_second': 561.495, 'eval_steps_per_second': 17.649, 'epoch': 8.0}
{'loss': 1.7332, 'grad_norm': 1.7297613620758057, 'learning_rate': 5.7035714285714295e-05, 'epoch': 8.99}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6407182216644287, 'eval_runtime': 4.6042, 'eval_samples_per_second': 559.71, 'eval_steps_per_second': 17.593, 'epoch': 9.0}
{'loss': 1.711, 'grad_norm': 1.8106145858764648, 'learning_rate': 5.187142857142857e-05, 'epoch': 9.99}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.631450891494751, 'eval_runtime': 4.601, 'eval_samples_per_second': 560.097, 'eval_steps_per_second': 17.605, 'epoch': 10.0}
{'loss': 1.7036, 'grad_norm': 1.807066798210144, 'learning_rate': 4.6700000000000003e-05, 'epoch': 10.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6555012464523315, 'eval_runtime': 4.6196, 'eval_samples_per_second': 557.836, 'eval_steps_per_second': 17.534, 'epoch': 11.0}
{'loss': 1.6925, 'grad_norm': 1.8824546337127686, 'learning_rate': 4.1528571428571425e-05, 'epoch': 11.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6308748722076416, 'eval_runtime': 4.6357, 'eval_samples_per_second': 555.899, 'eval_steps_per_second': 17.473, 'epoch': 12.0}
{'loss': 1.6831, 'grad_norm': 2.2657229900360107, 'learning_rate': 3.6357142857142854e-05, 'epoch': 12.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6302309036254883, 'eval_runtime': 4.6095, 'eval_samples_per_second': 559.068, 'eval_steps_per_second': 17.573, 'epoch': 13.0}
{'loss': 1.6634, 'grad_norm': 1.9457443952560425, 'learning_rate': 3.118571428571428e-05, 'epoch': 13.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6137545108795166, 'eval_runtime': 8.6492, 'eval_samples_per_second': 297.946, 'eval_steps_per_second': 9.365, 'epoch': 14.0}
{'loss': 1.6555, 'grad_norm': 1.8875569105148315, 'learning_rate': 2.601428571428571e-05, 'epoch': 14.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.5990443229675293, 'eval_runtime': 14.0697, 'eval_samples_per_second': 183.16, 'eval_steps_per_second': 5.757, 'epoch': 15.0}
{'loss': 1.6503, 'grad_norm': 1.8996754884719849, 'learning_rate': 2.085e-05, 'epoch': 15.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.6132607460021973, 'eval_runtime': 14.256, 'eval_samples_per_second': 180.766, 'eval_steps_per_second': 5.682, 'epoch': 16.0}
{'loss': 1.6365, 'grad_norm': 1.8573732376098633, 'learning_rate': 1.5678571428571427e-05, 'epoch': 16.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.5969972610473633, 'eval_runtime': 15.1124, 'eval_samples_per_second': 170.522, 'eval_steps_per_second': 5.36, 'epoch': 17.0}
{'loss': 1.6338, 'grad_norm': 1.926599383354187, 'learning_rate': 1.0507142857142858e-05, 'epoch': 17.98}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.5750062465667725, 'eval_runtime': 15.4109, 'eval_samples_per_second': 167.219, 'eval_steps_per_second': 5.256, 'epoch': 18.0}
{'loss': 1.6299, 'grad_norm': 1.8220282793045044, 'learning_rate': 5.342857142857143e-06, 'epoch': 18.97}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.5601774454116821, 'eval_runtime': 14.4095, 'eval_samples_per_second': 178.84, 'eval_steps_per_second': 5.621, 'epoch': 19.0}
{'loss': 1.6323, 'grad_norm': 1.874803066253662, 'learning_rate': 1.7142857142857143e-07, 'epoch': 19.97}


  0%|          | 0/81 [00:00<?, ?it/s]

{'eval_loss': 1.603448510169983, 'eval_runtime': 14.3776, 'eval_samples_per_second': 179.237, 'eval_steps_per_second': 5.634, 'epoch': 20.0}


Overwriting existing adapter 'mlm_union_all_G'.


{'train_runtime': 3032.3522, 'train_samples_per_second': 152.97, 'train_steps_per_second': 4.782, 'train_loss': 1.7649665003809436, 'epoch': 20.0}


TrainOutput(global_step=14500, training_loss=1.7649665003809436, metrics={'train_runtime': 3032.3522, 'train_samples_per_second': 152.97, 'train_steps_per_second': 4.782, 'train_loss': 1.7649665003809436, 'epoch': 20.0})

In [15]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/81 [00:00<?, ?it/s]

>>> Perplexity: 4.80


In [16]:
import config.config as config

trainer.model.save_adapter(f"{config.Config.ADAPTER_SAVE_PATH}/{adapter_name}", adapter_name)


: 

#### This is a report of mlm for goverment domain using whole genre as target and then splitt after tokenization with seed 42. I used computer metrix for preplexity and early stopping 
The results before:
>>> Perplexity: 16.83

The results after:
>>> Perplexity: 4.80